packages/eslint-config/rules/bestpractices.js

```diff
@@ -0,0 +1,542 @@
+
+
+    /**
+     * disallow extending native types
+     */
+    'no-extend-native': ['error'],
+
+
+    /**
+     * disallow unnecessary calls to .bind()
+     */
+    'no-extra-bind': ['error'],
+
+
+    /**
+     * disallow unnecessary labels
+     */
+    'no-extra-label': ['error'],
+
+
+    /**
+     * disallow fallthrough of case statements
+     */
+    'no-fallthrough': ['error'],
+
+
+    /**
+     * disallow leading or trailing decimal points in numeric literals
+     */
+    'no-floating-decimal': ['error'],
+
+
+    /**
+     * disallow assignments to native objects or read-only global variables
+     */
+    'no-global-assign': ['error'],
+
+
+    /**
+     * disallow shorthand type conversions
+     */
+    'no-implicit-coercion': ['error'],
+
+
+    /**
+     * disallow variable and function declarations in the global scope
+     */
+    'no-implicit-globals': ['error'],
+
+
+    /**
+     * disallow the use of eval()-like methods
+     */
+    'no-implied-eval': ['error'],
+
+
+    /**
+     * disallow this keywords outside of classes or class-like objects
+     */
+    'no-invalid-this': ['error'],
+
+
+    /**
+     * disallow the use of the __iterator__ property
+     */
+    'no-iterator': ['error'],
+
+
+    /**
+     * disallow labeled statements
+     */
+    'no-labels': ['error'],
+
+
+    /**
+     * disallow unnecessary nested blocks
+     */
+    'no-lone-blocks': ['error'],
+
+
+    /**
+     * disallow function declarations and expressions inside loop statements
+     */
+    'no-loop-func': ['error'],
+
+
+    /**
+     * disallow magic numbers
+     */
+    'no-magic-numbers': ['error', {
+      ignore: [
+        -1, // for decrement
+        0, // for explicit length checks
+        1, // for increment
+        2, // for halving values
+        24, // for time math
+        60, // for time math
+        100, // for currency math
+        1000, // for time math
+      ],
+      enforceConst: true,
+    }],
+
+
+    /**
+     * disallow multiple spaces
+     */
+    'no-multi-spaces': ['error', {
+      ignoreEOLComments: false,
+    }],
+
+
+    /**
+     * disallow multiline strings
+     */
+    'no-multi-str': ['error'],
+
+
+    /**
+     * disallow new operators outside of assignments or comparisons
+     */
+    'no-new': ['error'],
+
+
+    /**
+     * disallow new operators with the Function object
+     */
+    'no-new-func': ['error'],
+
+
+    /**
+     * disallow new operators with the String, Number, and Boolean objects
+     */
+    'no-new-wrappers': ['error'],
+
+
+    /**
+     * disallow octal literals
+     */
+    'no-octal': ['error'],
+
+
+    /**
+     * disallow octal escape sequences in string literals
+     */
+    'no-octal-escape': ['error'],
+
+
+    /**
+     * disallow reassigning function parameters
+     */
+    'no-param-reassign': ['error', {
+      props: true,
+      ignorePropertyModificationsForRegex: [
+        '^acc(umulator)?', // for reduce accumulators
```

Comment:
I generally have *no* exceptions to this rule, could you explain why these instances need to have one?

Reply:
Most of these are kinda self explanatory if you know their context, but I can go down the list.

* `acc(umulator)` - for `array.reduce()`. Treating accumulator as a mutable object is acceptable because it simplifies logic with little repercussions since anything that isn't returned by the reduce function gets ignored.
* `desc(riptor)` - for decorators and `Object.defineProperty()` where it is common to need to mutate the property's descriptor
* `req(uest)` - Both this and response below are for express.js projects where it is common to need to mutate both the request and response properties
* `res(ponse)` - see above
* `Component` - for React specifically. `Component` is often used to refer to the component class in decorators, or when passing around react components directly instead of their instances.
* `ctx` - for koa.js environments for the same reason as `req` and `res` above.
* `draftState` - this is for the `immer` library, where it is explicitly expected of you to mutate the provided parameter.
* `event` - Actually! this one could be removed. This exception was lifted from `eslint-config-airbnb` and we don't use it. the exception was made for legacy workspace where setting `event.returnValue` was a thing... 
* `target` - For the same reasoning as `Component`, `target` is often used as the parameter providing the class for class decorators.

Reply:
You understand that this rule does not prevent you from modifying properties on an object, correct? It just prevents reassignment, e.g  ctx = {}
e.g ctx.status = 400 is still accepted 

I fail to see you would ever want to reassign ctx

Reply:
https://clapton.dev/u/2002/19ae310b.png

Reply:
I see what's going on here, it's because you are setting props: true in this rule. I'd argue for just removing that.

Reply:
The current config elects to also error on property reassignment. we COULD remove these exceptions if we turn prop checks off, but I'm of the mind that we should avoid any mutation of function arguments unless we assure it's safe to do so, which is why we guard against it. Using designated names for cases where it is acceptable is good practice to ensure we don't accidentally reassign something that isn't expected to change.

Reply:
A property modification is **not** a reassignment, I don't know why that option is even part of this rule, it should be a separate one, and it's false by default for a reason. 
Preventing property modification makes **no** sense to me.

Reply:
it's not a reassignment of a parameter, no, however it is considered a side-effect which could result in unexpected behavior if the changed property is expected to be a different value in another part of your code or to a library you're using. One good example of this is React's `static getDerivedStateFromProps(props, state) {}` lifecycle method. One may assume it is safe here to assign properties of the `state` argument based on the values of `props` instead of returning your new state object like you're supposed to. Since react performs separate tasks based upon the return value of `gDSFP`, directly setting state instead of returning your new state will ultimately lead to errors.

Reply:
Fine, I will give in to this insanity, I need an exception for "connection" which is an alternative name for ctx some places in my code, and "client" which is used in the websocket server and like Koa it expects you to mutate that object to use it

Reply:
If you still disagree we can set props to false and I'll override it in my config. Ultimately we are dealing with 2 separate approaches to handling data so I kinda figured this one was gonna conflict XD

Reply:
I don't really do this either except in the edge cases above, I'm fine with doing this rule even though I don't 100% agree with the logic it's also not harmful to follow it so whatever

Reply:
as I noted in #rattech, This preference comes from Functional programming principles of immutable state. Since we generally follow FP on the website, mutating any object is considered a bad thing.

Reply:
I can accept this as long as the two exceptions I requested above are added